// Copyright 2016 Google Inc. All Rights Reserved.

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//     http://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.


syntax = "proto2";

package openhtf;

import "google/protobuf/descriptor.proto";

extend google.protobuf.EnumValueOptions {
  // Field for storing an ANSI Unit-of-Measurement Code on an enum.
  // http://www.unece.org/cefact/codesfortrade/codes_index.html
  optional string uom_code = 50508554;
  optional string uom_suffix = 66336003;
}


message Units {
  /*
    Units that can be used when specifying test parameters.

    When adding a new unit, first look at the ANSI standard @ go/ansiunits.  If
    your unit is there be sure to specify a uom_code using the code in the doc.
    If it's not in the doc then don't specify a uom_code.

    For suffix use either an SI suffix or something that people recognize since
    it's usually geared towards displaying the unit in the UI.
  */
  enum UnitCode {
    // Next tag = 82;
    // NONE is the default, and for parameters which have no unit set.
    NONE = 1;
    PERCENT = 2 [(uom_code) = "P1", (uom_suffix) = "%"];
    // NO_DIMENSION means that there are units set, but they cannot be
    // expressed by a known dimension (such as a ratio)
    NO_DIMENSION = 3 [(uom_code) = "NDL"];
    PIXEL = 4 [(uom_code) = "PX", (uom_suffix) = "px"];
    PIXEL_LEVEL = 5 [(uom_code) = "PXL"];

    // Angle
    ROTATIONS_PER_MINUTE = 7 [(uom_code) = "RPM", (uom_suffix) = "rpm"];
    DEGREE = 60 [(uom_code) = "DD", (uom_suffix) = "°"];
    RADIAN = 61 [(uom_code) = "C81", (uom_suffix) = "rad"];
    DEGREE_PER_SECOND = 62 [(uom_code) = "E96", (uom_suffix) = "°/s"];

    // Time
    SECOND = 10 [(uom_code) = "SEC", (uom_suffix) = "s"];
    MHZ = 11 [(uom_code) = "MHZ", (uom_suffix) = "MHz"];
    HERTZ = 12 [(uom_code) = "HTZ", (uom_suffix) = "Hz"];
    MICROSECOND = 13 [(uom_code) = "B98", (uom_suffix) = "µs"];

    // Distance
    MILLIMETER = 21 [(uom_code) = "MMT", (uom_suffix) = "mm"];
    CENTIMETER = 22 [(uom_code) = "LC", (uom_suffix) = "cm"];
    METER = 23 [(uom_code) = "MTR", (uom_suffix) = "m"];
    PER_METER = 24 [(uom_code) = "M0R", (uom_suffix) = "m⁻¹"];
    // Acceleration
    GRAVITY = 27 [(uom_code) = "K40", (uom_suffix) = "g"];
    // Volume
    MILLILITER = 25 [(uom_code) = "MLT", (uom_suffix) = "mL"];
    CUBIC_FOOT = 26 [(uom_code) = "FTQ", (uom_suffix) = "Ft³"];
    CUBIC_METER = 28 [(uom_code) = "MTQ", (uom_suffix) = "m³"];

    // Sound
    DECIBEL = 30 [(uom_code) = "2N", (uom_suffix) = "dB"];
    DECIBEL_MW = 31 [(uom_code) = "2N", (uom_suffix) = "dBmW"];

    // Electric Current
    MICROAMP = 32 [(uom_code) = "B84", (uom_suffix) = "µA"];
    MILLIAMP = 33 [(uom_code) = "4K", (uom_suffix) = "mA"];
    AMPERE = 29 [(uom_code) = "AMP", (uom_suffix) = "A"];
    // Electric Potential
    MICROVOLT = 34 [(uom_code) = "D82", (uom_suffix) = "µV"];
    MILLIVOLT = 38 [(uom_code) = "2Z", (uom_suffix) = "mV"];
    VOLT = 35 [(uom_code) = "VLT", (uom_suffix) = "V"];
    // Capacitance
    PICOFARAD = 36 [(uom_code) = "4T", (uom_suffix) = "pF"];
    // Charge
    COULOMB = 37 [(uom_code) = "COU", (uom_suffix) = "C"];
    // Power
    WATT = 39 [(uom_code) = "WTT", (uom_suffix) = "W"];

    // Temperature
    DEGREE_CELSIUS = 40 [(uom_code) = "CEL", (uom_suffix) = "°C"];
    KELVIN = 41 [(uom_code) = "KEL", (uom_suffix) = "K"];

    // Data
    BYTE = 50 [(uom_code) = "AD", (uom_suffix) = "B"];
    MEGA_BYTES_PER_SECOND = 51 [(uom_code) = "P95", (uom_suffix) = "MB/s"];

    // Force
    NEWTON = 70 [(uom_code) = "NEW", (uom_suffix) = "N"];

    // Volume Flow Rate
    CUBIC_CENTIMETER_PER_SEC = 80 [(uom_code) = "2J", (uom_suffix) = "cm³/s"];
    // Pressure
    MILLIBAR = 81 [(uom_code) = "MBR", (uom_suffix) = "mbar"];

    // Magnetism
    MICROTESLA = 90 [(uom_code) = "D81", (uom_suffix) = "µT"];
  }
}
